crates/merkle-inclusion/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+pub mod traits;
+
+use crate::traits::GadgetDigest;
+use bellpepper_core::boolean::Boolean;
+use bellpepper_core::{ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+
+// HashValue represents the digest type output from the hash function
+pub type HashValue = Vec<Boolean>;
+pub type Key = Vec<Boolean>;
+
+/// Represents a leaf node in a Jellyfish Merkle Tree.
+///
+/// Each leaf node contains a key-value pair, where the key is used to determine the position in the tree and the value
+/// is the data stored.
+#[derive(Debug, Clone)]
+pub struct Leaf {
+    key: Key,
+    value_hash: HashValue,
+}
+
+impl Leaf {
+    pub fn new(key: Key, value_hash: HashValue) -> Self {
+        Self { key, value_hash }
+    }
+    pub fn hash(&self) -> &HashValue {
+        &self.value_hash
+    }
+
+    pub fn key(&self) -> &HashValue {
+        &self.key
+    }
+}
+
+/// Contains a leaf node and sibling hashes required for proof verification in a Merkle Tree.
+///
+/// The `Proof` struct is essential for the Merkle Tree's cryptographic verification process, providing the necessary
+/// information to verify the integrity and authenticity of the data.
+#[derive(Debug, Clone)]
+pub struct Proof {
+    leaf: Leaf,
+    siblings: Vec<HashValue>,
+}
+
+impl Proof {
+    pub fn new(leaf: Leaf, siblings: Vec<HashValue>) -> Self {
+        Self { leaf, siblings }
+    }
+
+    pub fn leaf(&self) -> &Leaf {
+        &self.leaf
+    }
+
+    pub fn siblings(&self) -> &[Vec<Boolean>] {
+        &self.siblings
+    }
+}
+
+/// Verifies the Merkle Tree proof against an expected root hash.
+///
+/// This function reconstructs the root hash from the given leaf and sibling hashes and compares it against the provided
+/// expected root hash.
+///
+/// # Arguments
+/// * `cs` - A mutable reference to the constraint system.
+/// * `expected_root` - The expected root hash of the Merkle Tree.
+/// * `proof` - The proof containing the leaf and its sibling hashes.
+///
+/// # Returns
+/// A result containing the reconstructed root hash if successful, or a `SynthesisError` otherwise.
+pub fn verify_proof<E, CS, GD>(
+    mut cs: CS,
+    expected_root: Vec<Boolean>,
+    proof: Proof,
+) -> Result<Vec<Boolean>, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+    GD: GadgetDigest<E>,
+{
+    assert_eq!(expected_root.len(), GD::output_size() * 8);
+
+    assert!(
+        proof.siblings.len() <= proof.leaf().key().len(),
+        "Merkle Tree proof has more siblings ({}) than the key length ({}).",
+        proof.siblings.len(),
+        proof.leaf().key().len(),
+    );
+    //Assert that we do not have more siblings than the length of our hash (otherwise cannot know which path to go)
+
+    // Reconstruct the root hash from the leaf and sibling hashes
+    let mut actual_root_hash = proof.leaf().hash().to_vec();
+
+    let key_iterator =
+        conditional_reverse::<_, _, GD>(proof.leaf().key().iter().take(proof.siblings().len()));
```

Comment:
if I have for example :
- `proof.siblings.len() == 5` and 
- `proof.leaf().key().len() == 8`, and 
- `GD::is_little_endian() == true`

The sequence of indexes of `proof.leaf.key()` that will be used to thread `proof().siblings()` is `[4, 3, 2, 1, 0]`. Is that what the user should expect, or should they expect `[7, 6, 5, 4, 3]` instead? Is the MSB of the key not at index 7 in LE order?

I think there's two ways to go about this:
1. explicitly state everywhere that key and gadget should have the same endianness, and in that case, reverse first, and  `.take(proof.siblings.len())` after. This is closer to the Aptos spec (note the mention of MSB of the key there), but may be less simple for other Merkle algorithms.
2. not explicitly make any relation between the gadget's endianness and the key's iteration order, and say that in a proof, key and siblings should be iterable in *the same* order, whatever it is: to the point that the algorithm should operate on the iterator formed by `proof.leaf.key().iter().zip(proof.siblings.iter())` (after checking `proof.keaf.key().len() > proof.siblings().len()`)

(I don't particularly care which of these two solutions we use, but I am mindful that we put the user in a situation to know precisely which approach we're enforcing.)

Reply:
I've went along with solution 2. , I think it should suffice in our context.